### PR TITLE
Rosenpass API infrastructure

### PR DESCRIPTION
**Pull request by koraa**
This implements the necessary infrastructure for the Rosenpass API.

Commit history:
- A variety drive-by refactoring changes, code quality and developer experience enhancements
- Utilities needed for implementing the API
- The API Infrastructure itself

I am interested in your reviews: @pqcfox @blipp
You too if you feel up to it: @cvengler 

@wucke13 Could you take a look at the changes in the nix file? My goal was to enable rust analyzer as part of the dev setup so I could have decent CI support writing this in my editor.

For the record, this was a humongous effort; mostly to an absurd amount of boilerplate code required to accurately parse and serialize network messages:
- +18 files (16% increase)
- +438 empty lines of code (18% increase)
- +75 comment lines (4% increase)
- +2069 lines of code (12% increase)

We should later investigate how to slim a lot of this code down by using procedural macros.
There definitely is a potential to reduce the lines of code by between 25% and 50%.

### CODE REVIEW

**Comment by koraa** [comment]
I am aware of the missing tests, due to deadline constrains I won't be spending a ton of time adding the missing tests before merging this.

@blipp If you are interested in some good rust starter tasks, writing some of the missing unit tests and documentation would be appreciated :)

File: flake.nix
Comment (wucke13) [inline flake.nix:414]: I think it is fine. However, pulling in the full toolchain is fat, i.e. this includes:

- cargo
- cargo-clippy
- cargo-fmt
- cargo-miri
- clippy-driver
- miri
- rls
- rust-analyzer
- rust-gdb
- rust-gdbgui
- rust-lldb
- rustc
- rustdoc
- rustfmt

If we are fine with having a more fat devShell, then the nice thing is that it is really _complete_.

**Review by wucke13 — APPROVED** [review]

File: flake.nix
Comment (koraa) [reply flake.nix:414]: @wucke13 Thanks for the review!

File: util/src/length_prefix_encoding/decoder.rs
Comment (pqcfox) [inline util/src/length_prefix_encoding/decoder.rs:22]: [nit] typo: `lage` -> `large`

File: util/src/length_prefix_encoding/decoder.rs
Comment (pqcfox) [inline util/src/length_prefix_encoding/decoder.rs:135]: [nit] typo: `colect` -> `collect`

**Review by pqcfox — CHANGES_REQUESTED** [review]
Looks excellent! Found a few potential typos which I think are important to fix before merge, and a handful of nitpicks. Feel free to resolve the latter at your discretion if you think any of them aren't expedient.

**Comment by pqcfox** [comment]
Just pushed the last commit incorporating the above fixes--or rather, the ones I'm confident enough to make--to `dev/karo/api` on my fork of Rosenpass, since I don't seem to have commit access even though I'm in the RP org now.

Feel free to cherry-pick that commit over, or if there's a way to add me directly to the repo I can do so. LGTM once that's added, so I'll mark as approved now to not hold things up.

**Review by pqcfox — APPROVED** [review]

**Comment by blipp** [comment]
Sorry, my mistake, I thought that Write access is enough, but to push to pull requests, you need Maintain access. I gave this to you now, and I also pushed your commit to this pull request to unblock this faster.

**Pull request merged**
